### DISCUSSION TITLE
Private store references

### DISCUSSION THREAD

**Discussion by luisherranz** [WordPress/gutenberg#51596] [Interactivity API]
I've started thinking if we should expose a system to limit the access of some parts of the store to certain namespaces to avoid compatibility issues. For example, an e-commerce site may choose to expose `actions.coolCommerce.addToCart` publicly, but not `actions.coolCommerce.calculateTax`.

The places where we would have to limit access are:

- `wp_store` in PHP
- `store` in JS
- `data-wp-directive="ref"` refs during the PHP SSR
- `data-wp-directive="ref"` refs on the JS hydration

In terms of implementation:

- Filtering out the namespaces should be easy on `wp_store`. We could get the namespace from the block name declared on the `block.json`. I.e., `"name": "coolCommerce/someBlock"` -> `coolCommerce`.
- For the directives, we could populate the `data-wp-interactive` with the namespace, like `data-wp-interactive="coolCommerce"`. That would mean that `data-wp-interactive` won't be optional when client-side navigation is activated, but I think that's fine. The namespace would stay until another `data-wp-interactive` with a different namespace is found, similar to how `wp-context` works today.
- For JS, we can annotate the namespace of each directive when the `Directive` component is created using the value of the last `data-wp-interactive` as well.

The injection of `data-wp-interactive="namespace"` can be done automatically, of course.

In my opinion, we should not deviate much from the current developer experience, so the syntax for private references could be something as simple as starting the property with `_`. Something like:

```js
store({
  actions: {
    coolCommerce: {
      addToCart: () => {
        // Public.
      },
      _calculateTax: () => {
        // Private, only accessible by in the `coolCommerce` namespace.
      },
    },
  },
});
```

Once this is in place, we could use a system similar to the [lock/unlock](https://github.com/WordPress/gutenberg/tree/trunk/packages/private-apis) mechanism to grant access to a namespace internals:

```js
export const signCoolCommerce = __dangerousOptInToPrivateReferences(
  "I know using private references means my plugin will inevitably break on the next release.",
  "coolCommerce"
);
```

```js
store(
  {
    actions: {
      myPlugin: {
        _someAction: ({ actions }) => {
          const tax = actions.coolCommerce._calculateTax();
        },
      },
    },
  },
  [signCoolCommerce] // acknowledge that this will break your plugin
);
```

This system would only pass those private references to other private references, to avoid leaking private references to third consumers.

```js
store(
  {
    actions: {
      myPlugin: {
        someAction: ({ actions }) => {
          const tax = actions.coolCommerce._calculateTax(); // undefined!
        },
      },
    },
  },
  [signCoolCommerce]
);
```

Signing in the HTML could be done inlining the warning, like this:

```html
<div
  data-wp-interactive="myPlugin"
  data-wp-sign--cool-commerce="I know using private references means my plugin will break on the next release"
>
  <button data-wp-on--click="actions.coolCommerce._calculateTax">
    Calculate Tax
  </button>
</div>
```

Or with a private reference:

```html
<div data-wp-interactive="myPlugin" data-wp-sign="sign.myPlugin._coolCommerce">
  <button data-wp-on--click="actions.coolCommerce._calculateTax">
    Calculate Tax
  </button>
</div>
```

```js
store({
  sign: {
    myPlugin: {
      _coolCommerce:
        "I know using private references means my plugin will break on the next release",
    },
  },
});
```

Maybe with this last option, we can unify the signing of the directives and store into one, by making the `sign` property (or whatever name we choose) special for that purpose.

---

There is a way to avoid signing because namespaces are not enforced: a person that wants to access `actions.coolCommerce._calculateTax` could create a block that is named `coolCommerce/myExternalBlock` or add a filter just before the SSR and change the value of `data-wp-interactive="coolCommerce"`. Also, they could gain access without signing if they delete their `data-wp-interactive` attribute and they are inner blocks of an interactive block from `coolCommerce`.

If that is unacceptable, everybody would have to _sign_ when they want to access private references, even for their own namespace. But I don't like this option because if someone is deliberately trying to avoid the signing, it means they already know what it means, and this would just make the DX worse for everybody else.

---

These are just my first thoughts and I haven't done any proof of concept of the hypotheses. But happy to hear other opinions 🙂

**Comment by adamziel** [comment]
Do you want to enable plugin authors to create their own private APIs? `lock()` and `unlock()` prevent WordPress extenders from using internal WordPress APIs. They are not exposed as something extenders could use. Here's a few thoughts that may come handy when building that functionality, though:

Using the `_` prefix would work, but it is a bit "magical" and doesn't make it obvious there's a special meaning. Also, it could resemble the `__experimental` notation from Gutenberg and confuse developers. How about a named wrapper function?

```js
store({
  actions: {
    coolCommerce: {
      calculateTax: internal(() => {
        // Private, only accessible by in the `coolCommerce` namespace.
      }),
    },
  },
});
```

BTW is anything preventing me from providing a `coolCommerce` override in my plugin and getting access to these internal/private APIs?

## Private APIs provided by the interactivity API in WordPress core

If you want to restrict parts of the interactivity API for internal core use, you can do that with `lock/unlock`. Here's a few notes.

> In my opinion, we should not deviate much from the current developer experience, so the syntax for private references could be something as simple as starting the property with _

The legacy experience for experimental APIs is prefixing public APIs with `__experimental`. [This is now deprecated](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#legacy-experimental-apis):

> All in all, don’t use the __experimental prefix for new APIs. Use plugin-only APIs and private APIs instead.

Here's how you register a private action:

https://github.com/WordPress/gutenberg/blob/c0077f7ec5d398c9e5e4a6c0814d53284fb01f75/packages/commands/src/store/index.js#L24-L31

So, in your example, it could be:

```js
export const instance = store({
  actions: {
    core: {
      // Public
      insertBlock: () => {}
    },
  },
});

// Private
unlock( instance ).registerPrivateActions( {
    core: {
        changeInternalBlockOrder: () => {}
    }
} ); 
```

And then you would use it like:

```js
store(
  {
    actions: {
      myPlugin: {
        someAction: ({ actions }) => {
          const invalid = actions.core.changeInternalBlockOrder(); // undefined!

          const valid = unlock(actions).core.changeInternalBlockOrder(); // valid
        },
      },
    },
  },
  [signCoolCommerce]
);
```

The actual API could settle on `unlock(actions.core)` or some other flavor, but this is the basic idea for locking.

[Original lock/unlock PR](https://github.com/WordPress/gutenberg/pull/46131). [Coding Guidelines and code examples](https://developer.wordpress.org/block-editor/contributors/code/coding-guidelines/#legacy-experimental-apis-plugin-only-apis-and-private-apis). [Package docs](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-private-apis/).

**Reply by luisherranz** [reply]
> `unlock(actions).core.changeInternalBlockOrder();`

I like that syntax. Maybe combined with separate functions to register the store:

- `store({...})` for public properties
- `privateStore({...})` for private properties

I guess that API would make a decent developer experience…

> BTW is anything preventing me from providing a `coolCommerce` override in my plugin and getting access to these internal/private APIs?

That's basically the problem. We either:

- Find a way to prevent the use of namespaces by external people to your plugin (difficult or impossible).
- Force everyone to use `unlock` on all the private APIs, even on their own.
- Acknowledge that people can cheat the system on purpose and don't grant backward compatibility privileges to those who change their namespace to cheat.

**Reply by adamziel** [reply]
Or you could scope privacy per npm module and not per store namespace, e.g. by having a package-specific unexported version of `lock()` and `unlock()`

**Reply by luisherranz** [reply]
> > `unlock(actions).core.changeInternalBlockOrder();`

> I like that syntax. Maybe combined with separate functions to register the store:

Unfortunately, I forgot that we also need to use this syntax in the directives references, and:

- We don't want to use anything that resembles JS to avoid confusion.
- We don't want to use different syntaxes for JS/PHP and references.

So that kind of rules out `unlock(actions)` 😔

> Or you could scope privacy per npm module and not per store namespace, e.g. by having a package-specific unexported version of lock() and unlock()

What do you mean? How would npm modules be involved in this process?

**Comment by adamziel** [comment]
Also cc @jsnajdr @youknowriad @noisysocks

**Comment by jsnajdr** [comment]
Let me summarize how I understand this problem. Currently, WP private APIs allow a group of `@wordpress/*` packages, a group of "friends", to use C++ terminology, to access each other's private APIs. The group is defined in the `CORE_MODULES_USING_PRIVATE_APIS` array:

https://github.com/WordPress/gutenberg/blob/12a9749e4393ca6f7ca8c17a16f45684524c60f1/packages/private-apis/src/implementation.js#L12-L27

No other module can opt-in into this private group.

Now, the proposal is to allow creating new friend groups, for example a group of `@woocommerce/*` packages. They can share APIs like `calculateTax` within the group. Only another Woo package can call `calculateTax`.

Is that right?

**Reply by luisherranz** [reply]
> Now, the proposal is to allow creating new friend groups, for example a group of @woocommerce/* packages. They can share APIs like calculateTax within the group. Only another Woo package can call calculateTax.

Interactive blocks (directives) and stores (server and client) rather than packages. But yes.

The goal is to give tools to these interactive blocks to define what parts of their APIs (`state`, `context`, `actions`, etc…) are public and what parts are private, so they don't have to maintain backward compatibility of the private parts.

**Reply by jsnajdr** [reply]
I understand, partially :) To understand fully, could we come up with some more realistic example than `coolCommerce`?

For example, I'm familiar with the Jetpack Podcast Player block that could use this. There is a `media-source` store, and some blocks register things to play, and other blocks can control the playback. Like ▶️/⏸/⏪/⏩ buttons, or a "table of contents" block that has clickable links to particular timestamps. The store has actions like `registerMediaSource`, `playMediaSource`, `setMediaSourceCurrentTime`.

There is a group of blocks, say, `media-player`, `media-control` and `media-toc`, that cooperate together. If we don't want to expose some of this functionality to other blocks, would these three block form a private friend group? Or is the `media-source` store also a standalone unit that's a member of the group?

**Reply by luisherranz** [reply]
Yes, that's the main gist of it 🙂

> If we don't want to expose some of this functionality to other blocks, would these three block form a private friend group? Or is the media-source store also a standalone unit that's a member of the group?

Right now, everything belongs to a single store, so that would be a private group in the store. But one store or multiple stores is not the problem. The problem is how we limit the connection between the directives references (like `data-wp-on--click="actions.mediaSource.playMediaSource"` for example) to the actual code that is executed. The reference is a just string pointer, so even if we create multiple stores, we need a way to determine what HTML element has the right to access each store via one of these pointers.

**Reply by jsnajdr** [reply]
> Right now, everything belongs to a single store, so that would be a private group in the store.

Being a private method is however a relationship between the store and the clients that can call the private method. So the clients also need to be part of the private group. Putting just the store into the group is not enough.

The `data-wp-on--click` directive is a part of _something_ -- is it the saved markup of a particular block? And what is the best enclosing "unit" that owns the directive, and that can be a member of a private friends group? In core, the "unit" are `@wordpress/*` packages, but here, I guess it needs to be a block like `jetpack/media-player`?

**Reply by luisherranz** [reply]
Exactly, the `data-wp-on--click` is a directive that is added to the HTML. And that HTML belongs to a block (unless we also add support for PHP templates). That was the reason I was suggesting to _mark_ the block HTML with `data-wp-interactive="jetpack"`. But that is very easy to hack, as anybody can create a block and use the `jetpack` namespace if they want.

I guess there's always going to be a way to hack this system, so probably the decision here is how far we want to go to make it harder. With special emphasis on making it very hard to hack without noticing.

**Comment by luisherranz** [comment]
Talking out loud here, but I'd like to share my latest thoughts.

First, these are two goals I would like to accomplish:

- Only plugins that are supposed to be extended would need to use private APIs, which are a minority, so this shouldn't affect/complicate the non-private APIs.
- We should find a way to use the same syntax in JS/PHP as in the directive references.

I'm not totally convinced about this yet, but if [using `_prop` is not explicit enough](https://github.com/WordPress/gutenberg/discussions/51596#discussioncomment-6263437), I have started thinking if we should use a new store property called `private` and centralize everything there. Maybe it would be easier in terms of mental model and implementation. We could also use different functions to populate the store, like [I mentioned here](https://github.com/WordPress/gutenberg/discussions/51596#discussioncomment-6263882).

With this syntax, maybe it'd also make sense to reverse the namespaces, so we have `namespace.type.name` instead of `type.namespace.name`.

We would also need a private context directive. It could be `data-wp-private-context`. If we reverse the namespace and the type, maybe it'd even make sense to move the namespace to the directive itself: `data-wp-private-context--cool-commerce='{ ... }'`

I'm not sure if it would be a good idea because moving from kebab-case to camelCase is not straightforward, and it would mean kebab-case namespaces would not be allowed (because there's no way to differentiate between kebab and camel case ones: `cool-commerce -> coolCommerce` or `cool-commerce -> cool-commerce`?).

If we reverse namespace and type, maybe it would make more sense to absorb it as a `wp_store`/`store` function parameter (although we would need to check that we can extract the correct TypeScript type before commiting to it).

Finally, I've also been wondering if instead of using `data-wp-interactive="namespace"` to access the private APIs, we should use a special directive that grants access _per element_ (instead of per block) to avoid leaking the grants down.

This would fix the leaking problem, but if we want to save the unlocking signature in the private scope, we have a circular loop if people are forced to unlock their own namespaces.

- To access `coolCommerce` I need a reference to unlock it.
- The unlocking signature should be private and in my own namespace, so something like `private.myPlugin.unlock.coolCommerce`.
- To access `private.myPlugin` I also need a private unlocking signature and we have a loop.

The only way to avoid that would be to add the unlocking signature (`"I know that..."`) directly in the HTML. This is probably not a good idea, but for now, I haven't thought of any other solution.

With all these changes, the new API would be:

```php
<?php
wp_store( 'coolCommerce', array(
  'state' => array(
    'publicGlobalValue' => '...'
  )
));

wp_private_store( 'coolCommerce', array(
  'state' => array(
    'privateGlobalValue' => '...'
  )
))
?>

<div
  data-wp-context--cool-commerce='{ "publicLocalValue": "..." }'
  data-wp-private-context--cool-commerce='{ "privateLocalValue": "..." }'
>
  <button
    data-wp-on--click="coolCommerce.actions.somePublicAction"
  >
    Trigger some public action
  </button>
  <button
    data-wp-on--click="private.coolCommerce.actions.somePrivateAction"
    data-wp-unlock--cool-commerce="I know using private references from external plugins means my code will break on their next release"
  >
    Trigger some private action
  </button>
</div>
```

```js
import { store, privateStore } from "@wordpress/interactivity";

store("coolCommerce", {
  actions: {
    somePublicAction: ({ private, coolCommerce }) => {
      // It can access the public store.
      coolCommerce.context.publicLocalValue;
      coolCommerce.state.publicGlobalValue;
      // It can access the private store only when unlocked.
      private.coolCommerce.context.privateLocalValue;
      private.coolCommerce.state.privateGlobalValue;
    },
  },
  unlock: {
    coolCommerce:
      "I know using private references from external plugins means my code will break on their next release",
  },
});

privatestore("coolCommerce", {
  actions: {
    somePrivateAction: ({ private, coolCommerce }) => {
      // It can access the public store.
      coolCommerce.context.publicLocalValue;
      coolCommerce.state.publicGlobalValue;
      // It can access its own private store without being unlocked.
      private.coolCommerce.context.privateLocalValue;
      private.coolCommerce.state.privateGlobalValue;
      // It can access other private stores only when unlocked.
      private.otherPlugin.actions.otherPrivateAction();
    },
  },
  unlock: {
    otherPlugin:
      "I know using private references from external plugins means my code will break on their next release",
  },
});
```

---

Again, I'm not sure about any of these ideas. I'm just sharing my latest thoughts to keep the conversation going, and in case they trigger better ideas from someone else.

**Reply by jsnajdr** [reply]
Looking at your example, I have some vague ideas: first, the two `coolCommerce` stores, the public and private one, they are not implicitly related to each other, are they? What I mean is that they are two separate stores that just happen to share the same name, in different namespaces, but they could as well have _different_ names, right?

Second, the private store looks very much like a local variable. In the PHP template, the references to the private store are local, and also in the JS script file they are local. The JS file could look like:
```js
const privateCool = privatestore("coolCommerce_internal", {
  actions: {
    somePrivateAction: ({ coolCommerce }) => {
      coolCommerce.something; // reference to registered public store
      privateCool.something: // reference to local variable rather than some registry object
    }
  }
});

store("coolCommerce", {
  action: {
    somePublicAction: () => {
      privateCool.something; // reference to module-local variable
    }
  }
});
```

Alternatively, `privatestore()` could return an opaque identifier, used to access a public registry. You can get the private store from the public registry only if you hold the identifier.
```js
const opaqueCool = privatestore( ... );

somePublicAction = ( stores ) => {
  stores.get( opaqueCool ).somePrivateAction(); // Lookup in a WeakMap
}
```

So, the idea is that public stores are something like `export const publicStore = ...` in a JS module, while a private store is like `const privateStore = ...` local variable. Could it be helpful?

**Reply by luisherranz** [reply]
That's interesting, but the challenge is how to connect those with the HTML. Basically, how to give you access to those private methods while preventing _other_ people from using them.

```html
<!-- This HTML is in my block, it's fine! -->
<button data-wp-on--click="coolCommerce_internal.actions.somePrivateAction">
  Click me
</button>
```

```html
<!-- This HTML is in an external block, this is not fine! -->
<button data-wp-on--click="coolCommerce_internal.actions.somePrivateAction">
  Click me
</button>
```

**Comment by luisherranz** [comment]
I've got an idea that I think could work, based on the idea of using local variables to hide a value ([as Jarda pointed out here](https://github.com/WordPress/gutenberg/discussions/51596#discussioncomment-6301436)) and an idea to count the number of times the unlocking functions are called without using the unlocking signature.

- Plugin authors can unlock their own private stores using a function.
- They don't need to use the unlocking signature (`I know...`).
- That function returns a hash or id that is in their local scope and other plugins can't access.

```php
$unlock_hash = unlock_private_store( 'coolCommerce' );
```

The first protection mechanism is that this function can only be called once per private store namespace.

- If an external plugin calls it as well, WordPress outputs an error message.
- It doesn't matter whether the rightful owner or the external plugin calls it first, it will throw the second time it's called, preventing WordPress from running.

If an external plugin wants to get the hash, it can do so by passing the unlocking signature to the function, and therefore acknowledging that their plugin can break. When the unlocking signature is passed, the function doesn't throw even if it's called multiple times.

```php
$unlock_hash = unlock_private_store( 'coolCommerce', 'I know that my...' );
```

The second protection mechanism is that the hash needs to be present in the HTML tags where private directives are used.

```php
<?php
$unlock_hash = unlock_private_store( 'coolCommerce' );
$wrapper_attributes = get_block_wrapper_attributes();
?>

<div <?php echo $wrapper_attributes; ?>>
  <button
    data-wp-on--click="private.coolCommerce.actions.toggle"
    data-wp-unlock="<?php echo $unlock_hash; ?>"
  >
    <?php _e( 'Toggle' ); ?>
  </button>
</div>
```

This prevents anyone without a valid hash from accessing the private store in their directive references.

Internally, the `unlock_private_store` function stores a reference `hash` <-> `coolCommerce`. If the `WP_Directives_Processor` finds a directive that tries to access a private reference from the `coolCommerce` namespace and doesn't have a valid hash, it throws, preventing WordPress from running.

If we check the hashes in the server, we don't need to do it again in the client. So if the `WP_Directives_Processor` finds the correct hash, it can remove the `data-wp-unlock` directive from the HTML.

`unlock_private_store` can't be called twice for the same namespace, so if more than one block needs to access the private store, it needs to be moved to an external file.

```php
<?php
// my-plugin/unlock.php
$unlock_hash = unlock_private_store( 'coolCommerce' );
```

```php
<?php
// my-plugin/block-1/render.php
require_once '../unlock.php';
// ...
?>
<div ...
  <button
    data-wp-on--click="private.coolCommerce.actions.toggle"
    data-wp-unlock="<?php echo $unlock_hash; ?>"
  >
    ...
```

```php
<?php
// my-plugin/block-2/render.php
require_once '../unlock.php';
// ...
?>
<div ...
  <button
    data-wp-on--click="private.coolCommerce.actions.toggle"
    data-wp-unlock="<?php echo $unlock_hash; ?>"
  >
    ...
```

Finally, we need to unlock private stores in the client, so they can be used by other store references. For this, we can use a similar function:

```js
import { store, unlockPrivateStore } from "@wordpress/interactivity";

const unlockHash = unlockPrivateStore("coolCommerce");

privateStore("coolCommerce", {
  state: {
    somePrivateValue,
  },
});

store(
  "coolCommerce",
  {
    selectors: {
      somePublicValue: ({ private }) => {
        return process(private.coolCommerce.state.somePrivateValue);
      },
    },
  },
  [unlockHash]
);
```

Again, the second time that the `unlockPrivateStore` function is called for the same namespace without the unlocking signature, the app throws and stops working preventing external plugins from calling it without the unlocking signature.

External plugins can still get a valid hash by using the unlocking signature:

```js
const unlockHash = unlockPrivateStore("coolCommerce", "I know that ...");

store(
  "myExternalPlugin",
  {
    // I can use `private.coolCommerce` here
  },
  [unlockHash]
);
```

If plugins need to use private references in more than one block, they need to create an external file and import the hash from there:

```js
// my-plugin/unlock.js
import { unlockPrivateStore } from "@wordpress/interactivity";

export default unlockPrivateStore("coolCommerce");
```

```js
// my-plugin/block-1/view.js
import { store } from "@wordpress/interactivity";
import unlockHash from "../unlock.js";

store(
  {
    // ...
  },
  [unlockHash]
);
```

```js
// my-plugin/block-2/view.js
import { store } from "@wordpress/interactivity";
import unlockHash from "../unlock.js";

store(
  {
    // ...
  },
  [unlockHash]
);
```

That's basically the gist of it.

---

Of course, there are still ways to hack this system, but they need to be done deliberately:

- Use the `render_block` to inspect the rightful block and search for the hash in the HTML.
- Use the `unlock_private_store` with the unlocking signature to get a valid hash, then expose it as a global so it can be used by other plugins that have not signed it.
- For plugins where they need to share the hash by more than one block, find and require the `unlock.php` file of the rightful owner.

I also don't consider this syntax final. Maybe `data-wp-unlock` is not the best way to add the hash to the HTML, for example. But I'm more interested in finding an approach that works than in the syntax at this point.

I'll keep thinking about it for a while to see if this approach has any drawbacks, but I would also appreciate any thoughts 🙂

**Reply by luisherranz** [reply]
An upgrade to this idea.

Unlike namespaces, block types are unique and should be difficult to hack. The `WP_Directive_Processor` should be able to know which one is processing at a time. Taking that into account, maybe we can replace the need for the hash, the `data-wp-unlock` directive and the require of a single `unlock.php` file.

This `unlock_private_store` call would register that the current block type (`cool-commerce/some-block`) can access the private namespace of `coolCommerce`.

```php
// plugins/cool-commerce/some-block/render.php
unlock_private_store( 'coolCommerce' );
```

Now, the `WP_Directive_Processor` should allow access to `private.coolCommerce` to any directive as long as the current block type is `cool-commerce/some-block`, and it will throw for other block types.

_The `unlock_private_store` function can be called repeatedly many times as long as the current block type is the same, so this approach works even when called from the `render.php` file of a block that is included multiple times on the page._

When multiple block types need to access the same private namespace, we can use the same approach and allow a _single_ call to the `unlock_private_store` function:

```php
// plugins/cool-commerce/plugin.php
// Single call with all the block types. No other block types can be added later.
unlock_private_store( 'coolCommerce', array(
  'cool-commerce/some-block',
  'cool-commerce/other-block',
  'cool-commerce/yet-another-block',
))
```

Now, the `WP_Directive_Processor` should allow access to `private.coolCommerce` to any directive as long as the current block type belongs to those listed in the array.

And as in the previous case, if someone else tries to call `unlock_private_store` without the unlocking signature, WordPress will crash.

```php
// plugins/other-plugin/some-other-block/render.php
unlock_private_store( 'coolCommerce' ); // Throws!
unlock_private_store( 'coolCommerce', 'I know that...' ); // This is fine.
```

**Reply by luisherranz** [reply]
The `unlock_private_store` linked to block types idea also solves the static blocks problem, which was missing before.

**Comment by luisherranz** [comment]
Sharing the [third-party store dependency on lazy loaded blocks](https://github.com/WordPress/gutenberg/discussions/52723#discussioncomment-6476391) discussion here because it may be related to this problem.

**Comment by luisherranz** [comment]
@darerodz has started implementing [this version](https://github.com/WordPress/gutenberg/discussions/51596#discussioncomment-6346400), and today we reviewed the whole API because there are things that are not quite there yet.

### PHP: Instead of throwing, remove the directives from the HTML

First, we noticed that WordPress should never throw in the server. That means that throwing when there is a second call to `unlock_private_store` is not an option.

Instead of that, if there is a second call to `unlock_private_store`, we should remove **all the directives of that store** from the HTML.

Removing all the directives (even the legit ones) is important as an extra security measure in case no blocks from the owner plugin are present on the page.

For example, imagine there are two calls to `unlock_private_store( 'coolCommerce' )`. The first one is legit, and adds a cool commerce block type:

```php
// plugins/cool-commerce/plugin.php
unlock_private_store( 'coolCommerce', array(
  'cool-commerce/some-block',
));
```

The second one is not legit, and comes from another plugin:

```php
// plugins/other-plugin/plugin.php
unlock_private_store( 'coolCommerce', array(
  'other-plugin/other-block',
));
```

If the plugin's `unlock_private_store` is executed first and is able to add the `other-plugin/other-block` block type and `cool-commerce/some-block` is not present on the page, it will give the false impression that everything is fine.

### JS: Use a build hash instead of an import

If the Interactivity API starts using a system of global imports, like Import Maps or Module Federation, people won't be able to use it to create a separate chunk that exports the unlock hash because anyone could be able to access it:

```js
// coolCommerce/unlock.js
import { unlockStore } from "@wordpress/interactivity";

export default unlockStore("coolCommerce");
```

```js
// coolCommerce/some-block/view.js
import unlock from "coolCommerce/unlock";

store(
  "coolCommerce",
  {
    // ...
  },
  {
    unlock,
  }
);
```

```js
// other-plugin/other-block/view.js
import unlock from "coolCommerce/unlock";

const { state, actions } = store("coolCommerce", null, {
  unlock,
});
```

In this case, `coolCommerce/unlock` must be a separate chunk, but only accessible to other coolCommerce modules. That is how Webpack chunks work today, but if we move to Import Maps, it won't be intuitive to know that you can't use Import Maps for this use case, so we think we should find a different method.

An alternative solution that doesn't require an external chunk could be to inject a random string that changes on each build, for example, a timestamp.

Webpack:

```js
const webpack = require("webpack");

module.exports = {
  // other webpack config ...

  plugins: [
    new webpack.DefinePlugin({
      BUILD_TIMESTAMP: JSON.stringify(Date.now()),
    }),
  ],
};
```

The API would become an option that accepts a string:

```js
const { state, actions } = store(
  "coolCommerce",
  {
    /* config store... */
  },
  { private: BUILD_TIMESTAMP } // This will be replaced with a timestamp string
);
```

Other bundlers have similar define capabilities, for example, `esbuild` API is:

```js
import * as esbuild from "esbuild";

await esbuild.build({
  // other esbuild config...
  define: {
    BUILD_TIMESTAMP: JSON.stringify(Date.now()),
  },
});
```

We could provide something like that already configured in `wp-scripts`.

The `store` would only accept two strings:

- The string that was first passed for that namespace.
- The experimental string: `I know that...`

Third-party plugins can copy the current timestamp, but that would change when the user updates the original plugin. They would do this instead:

```js
const { state, actions } = store("coolCommerce", null, {
  private: "I know...",
});
```

Calling `store("coolCommerce")` with any other string will make it throw.

```js
const { state, actions } = store("coolCommerce", null, {
  private: "A different string", // Throws!!
});
```

This solution is also more performant because it doesn't require an extra blocking import.

A similar thing will happen to `wp_initial_state`, and although it's not as important as `store`, plugins that want to protect their initial state should be able to do it with a similar technique.

_`wp_initial_state` is the newer version of `wp_store`: [link](https://github.com/WordPress/gutenberg/discussions/53586)_

**Reply by adamziel** [reply]
@luisherranz this idea sounds similar to using secret key – @gziolo explored that in #41278. @peterwilsoncc [shared](https://github.com/WordPress/gutenberg/pull/41278#issuecomment-1145442652) why that wasn't a suitable solution for WordPress core:

>  * The environment variable will need to be included in the version hash
>    * this will modify the generated files each time the build is run and result in git/svn becoming dirty as a lot of these files are committed to WP-Develop
>    * The npm tests on WP-Develop will never pass [due to these lines](https://github.com/WordPress/wordpress-develop/blob/27963429a1d1a5f29e0bb7709bc2d49477323acb/.github/workflows/test-npm.yml#L87-L88).
> * The bundle for each WordPress update will need to include all the JavaScript files
>   * this caused problems for a minor release some time back (I'll research the exact version when I return from leave)
>   * bigger bundles increase the chance an update will fail

There's also some back&forth about using a stable hash in that thread.

**Reply by luisherranz** [reply]
I see. Thanks for the heads up, Adam 🙂

So even though a build-generated string is not suitable for WordPress/Gutenberg, I still think it's better to give it as an option to third-party plugins than having to always resort to the `"I know..."` string.

However, we would need a system to prevent store leakage when there's no build-generated string. I'd say we should simply use `private: true` for this:

```js
const { state, actions } = store(
  "coolCommerce",
  {
    /* config store... */
  },
  { private: true } // No string is stored. This can only be accessed with the "I know..." string from now on.
);
```

Plugins that only use the private store in a single file could also follow this approach.

And in WordPress/Gutenberg, we simply use the `"I know..."` string, like what you did in `@wordpress/private-apis`.

Thoughts? 🙂

**Reply by DAreRodz** [reply]
> Plugins that only use the private store in a single file could also follow this approach.

I think it would be a good idea for simple stores. 😄 In any case, we must remember to developers that they must not export their private stores, even for internal usage (or they will become public!).

**Reply by adamziel** [reply]
> So even though a build-generated string is not suitable for WordPress/Gutenberg, I still think it's better to give it as an option to third-party plugins 

@luisherranz Sure, that sounds reasonable!

I'd just make sure we keep the concepts of "WordPress internal" and "a store I can make private as a developer" separate. Let's not advertise the `lock()`/`unlock()` API or even tell the API consumers about the "I know..." strings.

For private stores in WordPress core, I'd stick to an established way of working with internal APIs, which is calling `lock` and `unlock`:

```js
export const privateApis = {};
const { state, actions } = store(
  "coolCommerce",
  {
    /* config store... */
  }, { private: true }
}
lock( privateApis, {
    store,
    actions
} );
```

And then we tell the bundler to never put build-time strings in `@wordpress` packages.

**Comment by luisherranz** [comment]
We should also add the possibility of having a read only store, where this is allowed:

```js
const { state, actions } = store("coolcommerce");
```

but this is not:

```js
const { state, actions } = store("coolcommerce", {
  // Add or overwrite parts of the store...
});
```

It should be easy to do now that we have the privacy the mechanism in place.

cc: @darerodz

**Comment by luisherranz** [comment]
The server-side part of my [last proposal](https://github.com/WordPress/gutenberg/discussions/51596#discussioncomment-7198628) is based on the assumption that we would know which part of the HTML belongs to which block during the processing, which is difficult because, after the block rendering, the block boundaries disappear.

```php
// plugins/cool-commerce/plugin.php
unlock_private_store( 'coolCommerce', array(
  'cool-commerce/some-block',
));
```

So we should search for a solution that doesn't require us to know that. This basically means that we need to move that information to the HTML. Of course, as the HTML can be modified by plugins, this method will be less secure, but hopefully it will be good enough to discourage people from hacking the system in weird ways instead of just using the `I know that my plugin will break...` method. So the goal is to come up with an API that makes things difficult enough, but it's not too cumbersome to use.

First, the most obvious solution would be to put the block name in the `data-wp-interactive` directive and keep the same `unlock_private_store` logic.

```html
<div data-wp-interactive='{ "blockName": "cool-commerce/some-block" }'></div>
```

The problems with this approach are:

1. It'd be very easy to add the block name of a legitimate block to your HTML and hack the system.
2. As we don't have block boundaries, the inner blocks of that block that don't have a `data-wp-interactive` directive will inherit the same `blockName`.

To solve 2., I don't think there's a better solution than instructing blocks using private stores to wrap their inner blocks with another directive to stop the leakage:

```php
<div data-wp-interactive='{ "blockName": "cool-commerce/some-block" }'>
  <h1>Some title</h1>
  <div data-wp-lock-private-stores>
    <?php echo $content; ?>
  </div>
</div>
```

Maybe one day we can automate this with the HTML API, but until that moment, devs will need to add it.

To solve 1., my current idea is to add an additional function that can be used in the `render_callback` function and the `render_block` filter to retrieve a single hash for a private store.

This would be the workflow:

- The `unlock_private_store` works in the same way as described before:
  - The plugin author uses the `unlock_private_store` function to add a list of allowed blocks:
    ```php
    // plugins/cool-commerce/plugin.php
    unlock_private_store( 'coolCommerce', array(
      'cool-commerce/some-block',
    ));
    ```
  - That function can be called only once per private store. If it's called more than once, it does nothing the second time (triggers `__doing_it_wrong`?). It can be called more than once if it includes the `I know that my plugin will...` or the same list of blocks.
    ```php
    // plugins/other-plugin/some-other-block/render.php
    unlock_private_store( 'coolCommerce', $more_blocks ); // Throws!
    unlock_private_store( 'coolCommerce', $more_blocks, 'I know that...' ); // This is fine.
    ```
- There is a new function that can only be used in the `render_callback` and `render_block` and returns a hash.
  - It could be a function:
    ```php
    <?php
    $hash = get_private_store_hash();
    ?>
    <div data-wp-interactive='{ "privateStores": "<?php echo $hash; ?>" }'></div>
    ```
  - But it could also be a directive directly:
    ```php
    <div <?php echo data_wp_unlock_private_stores(); ?>>...</div>
    ```
  - That function knows which block is being rendered, similar to what `WP_Block_Supports` is [doing here](https://github.com/WordPress/wordpress-develop/blob/cb85d888a0eee6f8d1636a4a8219127698a07423/src/wp-includes/class-wp-block.php#L256-L260).
  - If the block is not allowed, it does nothing (triggers `__doing_it_wrong`?).
- During the server directive processing:
  - Each time we process a private store, we store the hash.
  - If there is no hash, it removes the directives from the HTML (that's the protection).
  - If there is only one hash, it processes everything fine.
  - If it finds a second hash, it removes the directives from the HTML (that's the protection).

People trying to hack this system will have to inspect a legitimate block, find the hash in the HTML, then modify their own block and add the hash to it. Or if their block is an inner block of a legitimate block, modify the parent block to remove the `data-wp-lock-private-stores` directive.

Hopefully, these are cumbersome enough to discourage people and encourage them to use `unlock_private_store` with the `I know that...` string instead.

---

Thoughts? Any other ideas?
